### PR TITLE
AW--Implemented Index/Create/Edit Page for Rider Application from Member's View

### PR DESCRIPTION
**Pull request by Kanagawa-okiNamiUra**
In this PR, I implemented Index/Create/EditPage for Rider Application from Member's View (in the nav bar "Apply to be a rider." I also added a placeholder nav bar for admin's view.

Notice that this PR does not address not the fourth issue in the instruction because the show page, where fields not disabled in the form should be read only. I could not figure out how to do that other than writing a new Table component, which could be demanding considering the time I had. For now, the show button in the table will lead to the edit page, which is very similar to what it is supposed to look like.

Running Instance at: https://gride-albert-dev.dokku-03.cs.ucsb.edu/

An implementation note is that we (who have both ROLE_ADMIN and ROLE_MEMBER) will only have ROLE_MEMBER when we open the Index page. This is achieved by creating a copy of currentUser and modify the copy for easier reuse of the table component. In reality, it would not be a problem for users and we admins are barely affected. A picture explanation is attached below.
image
- Notice the Admin field disappears when we are at the URL /apply/rider.
image
- When we are no longer on that page, we're admins again. (The admins field appears in the picture above)

Some more screenshots of the app:
image
image

### CODE REVIEW

**Review by Danny-You — DISMISSED** [review]
LGTM! Ready to be reviewed by TA / Instructor

**Review by InariHimeko — DISMISSED** [review]
LGTM

File: frontend/src/main/components/Nav/AppNavbar.js [hunk 1]
```diff
@@ -97,6 +97,16 @@ export default function AppNavbar({ currentUser, systemInfo, doLogout, currentUr
...
                 )
               }
+              {
+                hasRole(currentUser, "ROLE_MEMBER") && (
+                  <Nav.Link as={Link} to="/apply/rider">Apply to be a Rider</Nav.Link>
+                )
+              }
+              {
+                hasRole(currentUser, "ROLE_ADMIN") && (
+                  <Nav.Link as={Link} to="/admin/applications/riders">Rider Applications</Nav.Link>
```
Comment (pconrad) [inline frontend/src/main/components/Nav/AppNavbar.js]: This page doesn't seem to have an  entry in `App.js`, so it doesn't seem to be accessible for the admins.

Is that intentional?  Did I overlook something? 

Or is this a bug that needs to be fixed?

**Review by pconrad — CHANGES_REQUESTED** [review]
I think I found a bug.   I could be mistaken. Can you take a look?

**Comment by pconrad** [comment]
I think I found a bug in this PR.  Can someone on the team take a look?

It doesn't have to be the original PR author.  Anyone can take a look at this and try to fix it (or explain why it's not a bug).

Specifically, if you look at App.js on this branch:

https://github.com/ucsb-cs156-m23/proj-gauchoride-m23-9am-3/blob/Albert-RiderApp_III/frontend/src/App.js

There is no mapping for the route /admin/applications/riders

However, the menu option "Rider Applications" in the Navbar links to that URL, which is "not found".

image

**Comment by Kanagawa-okiNamiUra** [comment]
Thanks for pointing that out. I  will implement the Index Page for Rider Application in the next PR.

**Review by lawrencewang30 — DISMISSED** [review]
Went on dokku instance and saw that the bug is now fixed. All other options are working as expected. Can merge to main whenever possible

File: frontend/src/tests/pages/RiderApplication/RiderApplicationIndexPageMember.test.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        fireEvent.click(cancelButton);
+
+        expect(mockConfirm).toHaveBeenCalledWith("Are you sure you want to cancel this application?\n\nClick 'OK' to confirm or 'Cancel' to keep your application active.");
+        
+        // Use setTimeout to introduce a delay before checking the updated status
+        setTimeout(() => {
+        // Verify that the status is updated to "cancelled"
+            expect(getByTestId(`${testId}-cell-row-2-col-status`)).toHaveTextContent("pending");
+            }, 1000); // Adjust the delay duration as needed
+
```
Comment (pconrad) [inline frontend/src/tests/pages/RiderApplication/RiderApplicationIndexPageMember.test.js]: NO!  This is not how you do it.  This makes what we call a "flaky test", one that sometimes works and sometimes doesn't.    

Resist the temptation to solve race conditions in tests with timers.

Here's probably what It should be:

```suggestion

await waitFor( ()=> {
 expect(getByTestId(`${testId}-cell-row-2-col-status`)).toHaveTextContent("pending");
            }
```

File: frontend/src/tests/pages/RiderApplication/RiderApplicationIndexPageMember.test.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        fireEvent.click(cancelButton);
+
+        expect(mockConfirm).toHaveBeenCalledWith("Are you sure you want to cancel this application?\n\nClick 'OK' to confirm or 'Cancel' to keep your application active.");
+        
+        // Use setTimeout to introduce a delay before checking the updated status
+        setTimeout(() => {
+            expect(getByTestId(`${testId}-cell-row-2-col-status`)).toHaveTextContent("cancelled");
+            }, 1000); 
+
+    });
```
Comment (pconrad) [inline frontend/src/tests/pages/RiderApplication/RiderApplicationIndexPageMember.test.js]: NO!  see fix below

**Review by pconrad — CHANGES_REQUESTED** [review]
ACK!  Timers in tests are a big no no.

I've outlined what is likely the proper way to do this below.

**Review by lawrencewang30 — APPROVED** [review]
Saw that the appropriate changes were made to the test files and controller file. Dokku instance still working as expected, can merge to main whenever possible

**Review by pconrad — APPROVED** [review]
LGTM!

**Pull request merged**